File: backend/app/schemas/resource_metrics.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


RESOURCE_METRICS_CONFIG_KEY = "resource_metrics_config"


class NpuMetricPoint(BaseModel):
    collected_at: datetime
    npu_utilization: float = 0
    npu_total: float = 0
    npu_used: float = 0
    npu_available: float = 0
    executing_pods_count: int = 0
    pr_count: int = 0
    top_pods: List[Dict[str, Any]] = Field(default_factory=list)


class ClusterNpuMetrics(BaseModel):
    cluster_id: int
    cluster_name: str
    metrics: List[NpuMetricPoint] = Field(default_factory=list)


class NpuMetricsResponse(BaseModel):
    clusters: List[ClusterNpuMetrics] = Field(default_factory=list)


class NodeMetricPoint(BaseModel):
    """单台机器单个时间点的指标"""
    collected_at: datetime
    npu_utilization: float = 0
    npu_total: float = 0
    npu_used: float = 0
    npu_available: float = 0
    cpu_utilization: float = 0
    memory_utilization: float = 0
    executing_pods_count: int = 0


class NodeSeries(BaseModel):
    """单台机器的趋势序列"""
    node_name: str
    metrics: List[NodeMetricPoint] = Field(default_factory=list)


class ClusterNodeMetrics(BaseModel):
    """单个集群下的机器指标集合"""
    cluster_id: int
    cluster_name: str
    nodes: List[NodeSeries] = Field(default_factory=list)


class NodeMetricsResponse(BaseModel):
    """机器维度指标查询响应"""
    clusters: List[ClusterNodeMetrics] = Field(default_factory=list)


class ResourceMetricsConfigResponse(BaseModel):
    interval_minutes: int = 1
    retention_days: int = 30


class ResourceMetricsConfigUpdate(BaseModel):
    interval_minutes: Optional[int] = Field(None, ge=1, le=60)
    retention_days: Optional[int] = Field(None, ge=1, le=365)

File: frontend/src/services/resourceMetrics.ts
import api from './api'

export interface NpuMetricPoint {
  collected_at: string
  npu_utilization: number
  npu_total: number
  npu_used: number
  npu_available: number
  executing_pods_count: number
  pr_count: number
  top_pods: TopPodInfo[]
}

export interface TopPodInfo {
  name: string
  namespace: string
  npu: number
  pr_number: number | null
  pr_url: string | null
  phase: string | null
}

export interface ClusterNpuMetrics {
  cluster_id: number
  cluster_name: string
  metrics: NpuMetricPoint[]
}

export interface NpuMetricsResponse {
  clusters: ClusterNpuMetrics[]
}

export interface NodeMetricPoint {
  collected_at: string
  npu_utilization: number
  npu_total: number
  npu_used: number
  npu_available: number
  cpu_utilization: number
  memory_utilization: number
  executing_pods_count: number
}

export interface NodeSeries {
  node_name: string
  metrics: NodeMetricPoint[]
}

export interface ClusterNodeMetrics {
  cluster_id: number
  cluster_name: string
  nodes: NodeSeries[]
}

export interface NodeMetricsResponse {
  clusters: ClusterNodeMetrics[]
}

export interface ResourceMetricsConfig {
  interval_minutes: number
  retention_days: number
}

export interface ResourceMetricsConfigUpdate {
  interval_minutes?: number
  retention_days?: number
}

export const getNpuMetrics = async (params: {
  cluster_ids?: number[]
  time_range?: string
  start_time?: string
  end_time?: string
} = {}) => {
  const searchParams = new URLSearchParams()
  if (params.time_range) {
    searchParams.append('time_range', params.time_range)
  }
  if (params.start_time) {
    searchParams.append('start_time', params.start_time)
  }
  if (params.end_time) {
    searchParams.append('end_time', params.end_time)
  }
  params.cluster_ids?.forEach(id => searchParams.append('cluster_ids', String(id)))
  const response = await api.get<NpuMetricsResponse>('/resource-dashboard/metrics/npu', {
    params: searchParams,
  })
  return response.data
}

export const getNodeMetrics = async (params: {
  cluster_ids?: number[]
  node_names?: string[]
  time_range?: string
  start_time?: string
  end_time?: string
} = {}) => {
  const searchParams = new URLSearchParams()
  if (params.time_range) {
    searchParams.append('time_range', params.time_range)
  }
  if (params.start_time) {
    searchParams.append('start_time', params.start_time)
  }
  if (params.end_time) {
    searchParams.append('end_time', params.end_time)
  }
  params.cluster_ids?.forEach(id => searchParams.append('cluster_ids', String(id)))
  params.node_names?.forEach(name => searchParams.append('node_names', name))
  const response = await api.get<NodeMetricsResponse>('/resource-dashboard/metrics/nodes', {
    params: searchParams,
  })
  return response.data
}

export const getResourceMetricsConfig = async () => {
  const response = await api.get<ResourceMetricsConfig>('/resource-dashboard/metrics/config')
  return response.data
}

export const updateResourceMetricsConfig = async (data: ResourceMetricsConfigUpdate) => {
  const response = await api.put<ResourceMetricsConfig>('/resource-dashboard/metrics/config', data)
  return response.data
}